New repo structure

jamwaffles opened the pull request:
Closes #9 

Adds a basic Justfile and CI config as a step towards #6 and #4, but does not address the points raised in those issues, so I'll leave them open for now.

jamwaffles commented:
I flew threw this over my lunch break, so do let me know if I missed anything silly/obvious.

rfuest commented on justfile:
```diff
@@ -0,0 +1,20 @@
+
+#----------
+# Building
+#----------
+
+build: build-next build-0-6
+
+# Build the simulator
+build-next:
+    #!/usr/bin/env bash
+    pushd ./eg-next
+    cargo build --examples
+    popd
+
+# Run cargo test in release mode
+build-0-6:
+    #!/usr/bin/env bash
+    pushd ./eg-0.6
+    cargo build --examples
+    popd
```
```suggestion
#----------
# Building
#----------

# Build the examples for all versions
build-all: (build "next") (build "0.6")

# Build the examples for the given version
build version:
    cd "eg-{{version}}" && cargo build --examples
```

The CI config will also need to be updated to call `build-all` if you decide to apply this suggestion.

rfuest commented on .circleci/config.yml:
```diff
@@ -0,0 +1,44 @@
+    steps:
+      - checkout
+      - restore_cache:
+          key: v1-{{ .Environment.CIRCLE_PROJECT_REPONAME }}-{{ .Environment.CIRCLE_JOB }}-{{ checksum "eg-0.6/Cargo.lock" }}-{{ checksum "eg-next/Cargo.lock" }}
```
Do we want to store the `Cargo.lock` for `eg-next` in the repo? I think this would cause the git dependency to be fixed to a single release until `cargo update` is called, which wouldn't be the expected behavior for a daily build of this repo.

jamwaffles replied:
That's a good point. I've left `Cargo.lock` both tracked by git and in this cache key, but I'm running `cargo update` before every build in CI. This invalidates the cache somewhat (until `cargo update` is performed by a developer and committed), but keeps things consistent between the other folders and this one.

I can add `eg-next/Cargo.lock` to `.gitignore` if you think that's a better solution though. I'm just wondering if it will hinder local dev much or not.

jamwaffles replied:
Never mind! That approach causes more problems than it solves... I've gitignored  `eg-next/Cargo.lock`. Let's see if that works...

rfuest reviewed (changes requested):
Some nits, but I like the solution otherwise.

jamwaffles commented:
I mean...

```
note: /usr/bin/ld: final link failed: No space left on device
```

CI is never easy is it!

jamwaffles commented:
Builds now take a while (5 mins ish) but no longer run out of memory and/or disk space.

rfuest reviewed (changes requested):
The `eg-next/Cargo.lock` is ignored but still included in the PR.

Should we convert the CI build to a matrix build to reduce runtime by running more jobs parallel?

rfuest commented on .circleci/config.yml:
```diff
@@ -0,0 +1,44 @@
+version: 2.1
+
+commands:
+  # Checkout repo, restore cache and update Cargo.lock
```
```suggestion
  # Checkout repo and restore cache
```

jamwaffles commented:
I've removed the lockfile and converted the build to a matrix build - I didn't realise Circle finally added them! (probably ages ago by this point...)

Builds are now down to less than 3m which is pretty alright.

rfuest reviewed (approved):
LGTM. Will you change the GH config to make the builds required to pass after the PR is merged?

jamwaffles commented:
I will indeed. Should we require all 4 to pass or just a couple? If the latter, which would you suggest?

rfuest commented:
IMO all should be required to pass. If we encounter problems with some builds in the future we can always disable them later.

jamwaffles commented:
Makes sense. I've marked all 4 builds as required now.
